posts/navarroaxel/14.public-key-with-openid-connect-jkws.md

```diff
@@ -0,0 +1,127 @@
+
+const printJwksUri = async (issuer) => {
+  const response = await fetch(`${issuer}/.well-known/openid-configuration`);
+  const {jwks_uri} = await response.json();
+  console.log(jwks_uri);
+};
+
+const token = '<your_jwt>';
+const {iss} = jwt.decode(token);
+printJwksUri(iss);
+```
+
+We have the JWKS URI programmatically in Node.js! 🥳
+
+## Verifying the token
+
+Now that we know enough about JWKS, we can write a Node.js code to validate an OpenID token that discovers the JWKS URI if you don't know where it is.
+
+```javascript
+const {promisify} = require('node:util');
+const jwt = require('jsonwebtoken');
+const jwksClient = require('jwks-rsa');
+
+const fetchJwksUri = async (issuer) => {
+  const response = await fetch(`${issuer}/.well-known/openid-configuration`);
+  const {jwks_uri} = await response.json();
+  return jwks_uri;
+};
+
+const getKey = (jwksUri) => (header, callback) => {
+  const client = jwksClient({jwksUri});
+  client.getSigningKey(header.kid, (err, key) => {
+    if (err) {
+      return callback(err);
+    }
+    callback(null, key.publicKey || key.rsaPublicKey);
+  });
+};
+
+/**
+ * Verify an OpenID Connect ID Token
+ * @param {string} token - The JWT Token to verify
+ */
+const verify = async token => {
+  const {iss: issuer} = jwt.decode(token);
+  const jwksUri = await fetchJwksUri(issuer);
+  return promisify(jwt.verify)(token, getKey(jwksUri));
+};
+
+const token = '<your_jwt>';
+verify(token)
+  .then(() => console.log('Token verified successfully.'))
+  .catch(console.error);
+```
+
+⚠️ Did you notice that I didn't check who is the issuer? This code will accept any of them, even a malicious one. 😨 To control this just accept issuers from an allowed list.
+
+```javascript
+const allowedIssuers = [
+  'https://login.microsoftonline.com/common/v2.0',
+  'https://sandrino.auth0.com',
+];
+
+const fetchJwksUri = async (issuer) => {
+  if (!allowedIssuers.includes(issuer)) {
+    throw new Error(`The issuer ${issuer} is not trusted here!`);
+  }
+  const response = await fetch(`${issuer}/.well-known/openid-configuration`);
+  const {jwks_uri} = await response.json();
+  return jwks_uri;
+};
+```
+
+With this little change you're safe now 🔒
+
+## Conclusion
+
+Like I mentioned in [this post](https://dev.to/cloudx/build-a-discoverable-change-password-form-3f9m), the well-known URI is a standard method to get information for specific features, and issuers should implement the  `/.well-known/openid-configuration` to get the integration easier for their users.
+
+🧠 OpenID Connect is a safe way to authenticate users, but you always have to verify the token’s integrity in the Back-End side and check if it was created by a trusted issuer.
+
+🔑 Remember that this scenario only works with JWKS (when the certificate is pre-distributed to the clients the JWT `header` has `x5t` instead of `kid`). You can find examples with `public.pem` files in the [`jsonwebtoken`](https://github.com/auth0/node-jsonwebtoken) package.
```

Comment:
```suggestion
🔑 Remember that this scenario only works with JWKS (when the certificate is pre-distributed to the clients the JWT `header` has [`x5t`](https://www.rfc-editor.org/rfc/rfc7515#section-4.1.7) instead of `kid`). You can find examples with `public.pem` files in the [`jsonwebtoken`](https://github.com/auth0/node-jsonwebtoken) package.
```